### PR TITLE
feat: lido oracle

### PR DESCRIPTION
**Pull request by ruvaag**
* implements lido oracle off `new/lido`
* addresses comments from #2 
* adds fork tests for lido oracle

### CODE REVIEW

File: src/lido/LidoOracle.sol [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+    /// @notice Returns the value of baseAmount of baseAsset in quoteAsset terms.
+    /// @param base The asset that the user needs to know the value for.
+    /// @param quote The asset in which the user needs to value the `base`.
+    /// @param baseAmount The amount of `base` that the user wants to know the value in `quote` for.
+    /// @return quoteAmount The amount of `quote` that has the same value as `baseAmount`.
+    function valueOf(address base, address quote, uint256 baseAmount) external view returns (uint256 quoteAmount) {
+        if (base == address(STETH) && quote == address(WSTETH)) {
+            // value of given stETH amount, in terms of wstETH
+            return STETH.getSharesByPooledEth(baseAmount);
```
Comment (alcueca) [inline src/lido/LidoOracle.sol:34]: This should be `WSTETH.getWstEthByStEth(baseAmount)`, I think.

File: src/lido/LidoOracle.sol [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        } else if (base == address(WSTETH) && quote == address(STETH)) {
+            // value of given wstETH amount, in terms of stETH
+            return STETH.getPooledEthByShares(baseAmount);
```
Comment (alcueca) [inline src/lido/LidoOracle.sol:37]: And then this should be `WSTETH.getStEthByWstEth(baseAmount)`.

File: src/lido/LidoOracle.sol [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    /// @notice Returns the price of baseAsset in quoteAsset terms.
+    /// @param base The asset that the user needs to know the price for.
+    /// @param quote The asset in which the user needs to price the `base`.
+    /// @return baseQuotePrice The value of a minimum representable unit of `base` in `quote` terms, as an FP18.
+    function priceOf(address base, address quote) external view returns (uint256 baseQuotePrice) {
+        if (base == address(STETH) && quote == address(WSTETH)) {
+            // value of one stETH whole unit, in terms of wstETH
+            return (STETH.getSharesByPooledEth(STETH_SCALAR) * 1e18) / WSTETH_SCALAR;
```
Comment (alcueca) [inline src/lido/LidoOracle.sol:51]: Here it would be `WSTETH.tokensPerStEth() * 1e18 / WSTETH_SCALAR`.

File: src/lido/LidoOracle.sol [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        } else if (base == address(WSTETH) && quote == address(STETH)) {
+            // value of one wstETH whole unit, in terms of stETH
+            return (STETH.getPooledEthByShares(WSTETH_SCALAR) * 1e18) / STETH_SCALAR;
```
Comment (alcueca) [inline src/lido/LidoOracle.sol:54]: Here it would be `WSTETH.stEthPerToken() * 1e18 / STETH_SCALAR`.

**Review by alcueca — CHANGES_REQUESTED** [review]
I think that the contract that serves as source is WSTETH, instead of STETH (remember that this oracle returns the WSTETH/STETH and STETH/WSTETH feeds).

File: src/lido/LidoOracle.sol (hunk 1 above)
Comment (ruvaag) [reply src/lido/LidoOracle.sol:34]: it's a cheaper way to do the same, `WSTETH.getWstEthByStEth` just calls `STETH.getSharesByPooledEth` internally

refer: [@totomanov's comment](https://github.com/alcueca/oracles/pull/2#issuecomment-2016760007)

same for the corresponding `priceOf` call

File: src/lido/LidoOracle.sol (hunk 2 above)
Comment (ruvaag) [reply src/lido/LidoOracle.sol:37]: it's a cheaper alternative to do the same, `WSTETH.getStEthByWstEth` just calls `STETH. getPooledEthByShares` internally

refer: [@totomanov's comment](https://github.com/alcueca/oracles/pull/2#issuecomment-2016760007)

same for the corresponding priceOf call

**Comment by ruvaag** [comment]
> I think that the contract that serves as source is WSTETH, instead of STETH (remember that this oracle returns the WSTETH/STETH and STETH/WSTETH feeds).

happy to revert for clarity since the change is only semantic.

but looking at it as a WSTETH-STETH oracle (since it prices the WSTETH/STETH and STETH/WSTETH pairs) it feels appropriate to call the stETH function, which is marginally cheaper

File: src/lido/LidoOracle.sol (hunk 1 above)
Comment (alcueca) [reply src/lido/LidoOracle.sol:34]: You are right, I misread his comment

**Pull request merged**